lib/nerves_runtime/log_tailer.ex

```diff
@@ -0,0 +1,136 @@
+defmodule Nerves.Runtime.LogTailer do
+  use GenServer
+
+  require Logger
+
+  defp gen_server_name(:syslog), do: __MODULE__.Syslog
+  defp gen_server_name(:kmsg), do: __MODULE__.Kmsg
+
+  @doc """
+  `type` must be `:syslog` or `:kmsg` to indicate which log to tail with this
+  process. They're manages by separate processes both to isolate failures and
+  to simplify the handling of messages being sent back from the ports.
+  """
+  @spec start_link(:syslog | :kmsg) :: {:ok, pid()}
+  def start_link(type) do
+    GenServer.start_link(__MODULE__, type, name: gen_server_name(type))
+  end
+
+  @spec init(:syslog | :kmsg) :: {:ok, %{type: atom(), port: port(), buffer: binary()}}
+  def init(type) do
+    if Application.get_env(:nerves_runtime, :disable_syslog, false) do
+      :ignore
+    else
+      port =
+        type
+        |> port_binary_name()
+        |> open_port()
+
+      {:ok, %{type: type, port: port, buffer: ""}}
+    end
+  end
+
+  def handle_info({port, {:data, {:noeol, fragment}}}, %{port: port, buffer: buffer} = state) do
+    {:noreply, %{state | buffer: buffer <> fragment}}
+  end
+
+  def handle_info({port, {:data, {:eol, fragment}}}, %{type: type, port: port, buffer: buffer} = state) do
+    handle_message(type, buffer <> fragment)
+    {:noreply, %{state | buffer: ""}}
+  end
+
+  defp port_binary_name(:syslog), do: "log_tailer"
+  defp port_binary_name(:kmsg), do: "kmsg_tailer"
+
+  defp open_port(name) do
+    Port.open({:spawn_executable, executable(name)}, [
+      {:line, 1024},
+      :use_stdio,
+      :binary,
+      :exit_status
+    ])
+  end
+
+  defp executable(name) do
+    :nerves_runtime
+    |> :code.priv_dir()
+    |> Path.join(name)
+  end
+
+  defp handle_message(type, data) do
+    # Let's just use our own timestamp instead of trying to parse the one from the syslog message
+    timestamp = DateTime.utc_now() |> DateTime.to_iso8601()
```

Comment:
Can we just drop the timestamp completely? `Logger` already grabs a timestamp by default: https://github.com/elixir-lang/elixir/blob/v1.6.0/lib/logger/lib/logger.ex#L619.

Reply:
Ah yeah that should be sufficient; no need to do the same thing twice. I kinda wish it showed up as metadata for easier parsing, but it's pretty common to just have a bare timestamp be the first thing in a log message (e.g. this tends to be assumed by systems like Splunk).